numpy/_core/src/multiarray/descriptor.c

```diff
@@ -513,6 +514,11 @@ _convert_from_array_descr(PyObject *obj, int align)
                     "Field elements must be tuples with at most 3 elements, got '%R'", item);
             goto fail;
         }
+        if (PyObject_IsInstance((PyObject *)conv, (PyObject *)&PyArray_StringDType)) {
+            PyErr_Format(PyExc_TypeError,
+                         "StringDType is not currently supported for structured dtype fields.");
+            goto fail;
```

Comment:
I have no idea how complicated it is to add. It might not be so bad, I just really didn't want to dive into the code for handling structured dtypes for a feature that most people won't care about much.

It's also somewhat of a weird thing to do, since the string data won't actually be in the structured array, which presumably you're using for serialization or memory-mapping purposes or something like that.

Reply:
Also, if I loosen this restriction, right now users would quickly hit seg faults, so this prevents that from happening until we can come back and make something nicer happen.

Your and Matti's point at the meeting the other day made me think a bit about maybe adding support for a dtype like "T[8]" or something like that, where we re-use the stringdtype code but *always* use the short string optimization with a user-specified width, and maybe set `elsize` a little higher than that so the string length can fit.  *that* dtype would be a much better fit for use in structured arrays.

Reply:
From left-field, but still: we do allow `object` dtype in structured arrays, but definitely that makes things harder. "not currently supported" seems the right call!